src/Pim/Component/Catalog/Manager/ProductUniqueDataSynchronizer.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace Pim\Component\Catalog\Manager;
+
+use Doctrine\Common\Collections\Collection;
+use Pim\Component\Catalog\Model\AttributeInterface;
+use Pim\Component\Catalog\Model\ProductInterface;
+use Pim\Component\Catalog\Model\ProductUniqueData;
+use Pim\Component\Catalog\Model\ProductUniqueDataInterface;
+use Pim\Component\Catalog\Model\ProductValueInterface;
+
+/**
+ * Synchronize the $uniqueData persistent collection of the product with the unique values of the product.
+ * Those unique values come from the $values collection
+ * {@see Pim\Component\Catalog\Model\ProductValueCollectionInterface}.
+ *
+ * The only aim of the $uniqueData collection is to be able to save these information in the database via Doctrine.
+ *
+ * @author    Julien Janvier <julien.janvier@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ProductUniqueDataSynchronizer
+{
+    /**
+     * @param ProductInterface $product
+     */
+    public function synchronize(ProductInterface $product)
+    {
+        $productUniqueData = $product->getUniqueData();
+
+        foreach ($product->getValues()->getUniqueValues() as $value) {
+            $attribute = $value->getAttribute();
+
+            $uniqueData = $this->getUniqueDataFromCollection($productUniqueData, $attribute);
+            if (null !== $uniqueData) {
+                $uniqueData->setProductValue($value);
+            }
+            else {
+                $uniqueData = $this->createUniqueData($product, $value);
+                $product->addUniqueData($uniqueData);
+            }
+        }
+    }
+
+    /**
+     * @param Collection         $uniqueData
+     * @param AttributeInterface $attribute
+     *
+     * @return ProductUniqueDataInterface|null
+     */
+    protected function getUniqueDataFromCollection(Collection $uniqueData, AttributeInterface $attribute)
+    {
+        foreach ($uniqueData as $oneUniqueData) {
+            if ($attribute === $oneUniqueData->getAttribute()) {
+                return $oneUniqueData;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * @param ProductInterface      $product
+     * @param ProductValueInterface $value
+     *
+     * @return ProductUniqueDataInterface
+     */
+    protected function createUniqueData(ProductInterface $product, ProductValueInterface $value)
+    {
+        //todo: factory here?
```

Comment:
:+1:
I think the moment we have a `new` (except in a constructor maybe), even if very simple like here, it should be done in a factory.